lib/tbot/bot/bot.go

```diff
@@ -0,0 +1,371 @@
+/*
+ * Teleport
+ * Copyright (C) 2025  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package bot
+
+import (
+	"context"
+	"time"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/trace"
+	"go.opentelemetry.io/otel"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/gravitational/teleport/api/client/webclient"
+	machineidv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/machineid/v1"
+	apitracing "github.com/gravitational/teleport/api/observability/tracing"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/reversetunnelclient"
+	"github.com/gravitational/teleport/lib/tbot/bot/connection"
+	"github.com/gravitational/teleport/lib/tbot/client"
+	"github.com/gravitational/teleport/lib/tbot/internal"
+	"github.com/gravitational/teleport/lib/tbot/internal/carotation"
+	"github.com/gravitational/teleport/lib/tbot/internal/heartbeat"
+	"github.com/gravitational/teleport/lib/tbot/internal/identity"
+	"github.com/gravitational/teleport/lib/tbot/readyz"
+)
+
+var tracer = otel.Tracer("github.com/gravitational/teleport/lib/tbot/bot")
+
+// Bot runs a collection of services/outputs to generate and renew credentials
+// on behalf of non-human actors (i.e. machines and workloads).
+type Bot struct{ cfg Config }
+
+// New creates a Bot with the given configuration. Call Run to run the bot in
+// long-running "daemon" mode, or OneShot to generate outputs once and then exit.
+func New(cfg Config) (*Bot, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &Bot{cfg: cfg}, nil
+}
+
+// Run the bot until the given context is canceled.
+func (b *Bot) Run(ctx context.Context) (err error) {
+	ctx, span := tracer.Start(ctx, "Bot/Run")
+	defer func() { apitracing.EndSpan(span, err) }()
+
+	services, closer, err := b.buildServices(ctx)
+	if err != nil {
+		closer()
+		return trace.Wrap(err)
+	}
+	defer closer()
```

Comment:
Nit: if you deferred `closer()` right after the `buildServices` call, wouldn't it also apply to the error case above?